lib/fileutils.rb

```diff
@@ -8,6 +8,85 @@
 
 # Namespace for file utility methods for copying, moving, removing, etc.
 #
+# == What's Here
+#
+# First, what’s elsewhere. \Module \FileUtils:
+#
+# - Inherits from {class Object}[https://docs.ruby-lang.org/en/master/Object.html].
+# - Supplements {class File}[https://docs.ruby-lang.org/en/master/File.html]
+#   (but is not included or extended there).
+#
+# Here, module \FileUtils provides methods that are useful for:
+#
+# - {Creating}[rdoc-ref:FileUtils@Creating].
+# - {Deleting}[rdoc-ref:FileUtils@Deleting].
+# - {Querying}[rdoc-ref:FileUtils@Querying].
+# - {Setting}[rdoc-ref:FileUtils@Setting].
+# - {Comparing}[rdoc-ref:FileUtils@Comparing].
+# - {Copying}[rdoc-ref:FileUtils@Copying].
+# - {Moving}[rdoc-ref:FileUtils@Copying].
+#
+# === Creating
+#
+# - ::mkdir: Creates directories.
+# - ::mkdir_p, ::makedirs, ::mkpath: Creates directories,
+#   also creating ancestor directories as needed.
+# - ::link_entry: Creates a hard link.
+# - ::ln, ::link: Creates hard links.
+# - ::ln_s, ::symlink: Creates symbolic links.
+# - ::ln_sf: Creates symbolic links, overwriting if necessary.
+#
+# === Deleting
+#
+# - ::remove_dir: Removes a directory and its descendants.
+# - ::remove_entry: Removes an entry, including its descendants if it is a directory.
+# - ::remove_entry_secure: Like ::remove_entry, but removes securely.
+# - ::remove_file: Removes a file entry.
+# - ::rm, ::remove: Removes entries.
+# - ::rm_f, ::safe_unlink: Like ::rm, but removes forcibly.
+# - ::rm_r: Removes entries and their descendants.
+# - ::rm_rf, ::rmtree: Like ::rm_r, but removes forcibly.
+# - ::rmdir: Removes directories and their descendants.
+#
+# === Querying
+#
+# - ::collect_method: Returns the names of methods that accept a given option.
+# - ::commands: Returns the names of methods that accept options.
+# - ::have_option?: Returns whether a given method accepts a given option.
+# - ::options: Returns all option names.
+# - ::options_of: Returns the names of the options for a given method.
+# - ::pwd, ::getwd: Returns the path to the working directory.
+# - ::uptodate?: Returns whether a given entry is newer than given other entries.
```

Comment:
I think everything except `pwd` and `uptodate?` should be moved to a different section (perhaps called "Metadata", or a better name if you can think of one) since they're information about the methods in this class.

Only `pwd` here is actually related to querying file operations.

I think `uptodate?` should be in the "Comparing" section since it's comparing timestamps of files.

Reply:
I disagree.  I have not claimed that all these query the file system, but only that they are queries (i.e., return information, rather than making changes).

I don't think it's a stretch to view a method that returns a boolean as a query.

Reply:
> I disagree. I have not claimed that all these query the file system, but only that they are queries (i.e., return information, rather than making changes).

I realize that the docs doesn't claim to query the file system, but everything else are file system related except these meta methods, so it may be a bit confusing for the reader.

Tangentially related, I don't understand the use case of these meta-querying methods. I can't see why a user has to check what keyword arguments are accepted by a method or what method accepts a particular keyword argument.

> I don't think it's a stretch to view a method that returns a boolean as a query.

I think I would define a query as something that returns information about the file. For example, checking if a file exists (which returns a boolean) is a query. A comparison is taking two files and comparing their data, returning true or false. For example, checking if two files contains the same contents is a comparison. I think comparing metadata (which is what `uptodate?` is doing) falls under comparison.

Reply:
I don't see the use case either.

I''ve named the new section "Options" even though it's not a gerund.

Can you live with ::uptodate? remaining in Querying, so ::pwd won't be lonely?